sdk/identity/azure-identity/azure/identity/_credentials/certificate.py

```diff
@@ -76,6 +77,42 @@ def extract_cert_chain(pem_bytes):
     return b"".join(chain.splitlines())
 
 
+_Cert = NamedTuple("_Cert", [("pem_bytes", bytes), ("private_key", "Any"), ("fingerprint", bytes)])
+
+
+def load_pem_certificate(certificate_data, password):
+    # type: (bytes, Optional[bytes]) -> _Cert
+    private_key = serialization.load_pem_private_key(certificate_data, password, backend=default_backend())
+    cert = x509.load_pem_x509_certificate(certificate_data, default_backend())
+    fingerprint = cert.fingerprint(hashes.SHA1())  # nosec
+    return _Cert(certificate_data, private_key, fingerprint)
+
+
+def load_pkcs12_certificate(certificate_data, password):
+    # type: (bytes, Optional[bytes]) -> _Cert
+    from cryptography.hazmat.primitives.serialization import Encoding, NoEncryption, pkcs12, PrivateFormat
+
+    private_key, cert, additional_certs = pkcs12.load_key_and_certificates(
+        certificate_data, password, backend=default_backend()
+    )
+    if not private_key:
+        raise ValueError("The certificate must include its private key")
+    if not cert:
+        # mentioning PEM here because we raise this error when certificate_data is garbage
+        raise ValueError("Failed to deserialize certificate in PEM or PKCS12 format")
+
+    # This serializes the private key without any encryption it may have had. Doing so doesn't violate security
+    # boundaries because this representation of the key is kept in memory. We already have the key and its
+    # password, if any, in memory.
+    key_bytes = private_key.private_bytes(Encoding.PEM, PrivateFormat.PKCS8, NoEncryption())
+    pem_sections = [key_bytes] + [c.public_bytes(Encoding.PEM) for c in [cert] + additional_certs]
+    pem_bytes = b"".join(pem_sections)
+
+    fingerprint = cert.fingerprint(hashes.SHA1())  # nosec
```

Comment:
The tool is bandit. Its warnings are rare enough that I've never run it locally. It runs as a stage of the Analyze job, so I expect a bandit warning to fail a CI run. You suppress the warning when you really need to do something bandit considers insecure and have confirmed there is no danger. For example, the issue here is that a collision attack against SHA1 is feasible. In this context that means an attacker could create a forged cert having the same fingerprint as a legitimate one. That's okay because AAD has a copy of the legitimate cert and uses the fingerprint only to identify that copy. An attacker can't authenticate with such a forged cert because its signatures won't verify against the legitimate cert.